### ISSUE TITLE
[BUG/ISSUE]  GCC 9.2.1 compiler error (identified by the new CI build matrix)

### ISSUE THREAD

**Issue by LiamBindle** [geoschem/geos-chem#113]
Hi everyone,

The test CI build matrix pipeline that I set up for my GC-Classic fork identified that the GCC9 build fails. Below is a screenshot of the pipeline's results (see the actual pipeline [here](https://dev.azure.com/lrbindle/geos-chem/_build/results?buildId=153)). See my PR (#112) for adding CI pipelines to GEOS-Chem.

![Screenshot from 2019-10-24 14-28-03](https://user-images.githubusercontent.com/12772306/67520789-27bcd400-f66f-11e9-9a72-88d7b9ffab27.png)

Clicking "Job ubuntu_gcc9" gives a deeper look at what happened:

![Screenshot from 2019-10-24 14-28-20](https://user-images.githubusercontent.com/12772306/67520798-2d1a1e80-f66f-11e9-8c1e-ef8416db57cb.png)

Anyways, I thought this was a good demonstration of the build matrix, but this is an actual problem in GEOS-Chem 12.6.0. Here's the compiler error:

```
[ 50%] Building Fortran object HEMCO/Core/CMakeFiles/HCO.dir/hco_config_mod.F90.o
/__w/1/s/HEMCO/Core/hco_config_mod.F90:998:51:

  746 |           DO N = 1, nTags
      |                         2                          
......
  998 |                                         SplitInts, N, RC )
      |                                                   1
Error: Variable 'n' at (1) set to undefined value inside loop  beginning at (2) as INTENT(OUT) argument to subroutine 'hco_charsplit'
HEMCO/Core/CMakeFiles/HCO.dir/build.make:158: recipe for target 'HEMCO/Core/CMakeFiles/HCO.dir/hco_config_mod.F90.o' failed
make[3]: *** [HEMCO/Core/CMakeFiles/HCO.dir/hco_config_mod.F90.o] Error 1
make[2]: *** [HEMCO/Core/CMakeFiles/HCO.dir/hco_config_mod.F90.o.provides] Error 2
make[1]: *** [HEMCO/Core/CMakeFiles/HCO.dir/all] Error 2
make[1]: *** Waiting for unfinished jobs....
```

Note that the build matrix only tests the CMake build, so I don't know if this error is specific to CMake or something wider. 

Has anyone seen this before? Has anyone tried the 12.6.0 GNU Make build with GCC 9?

Liam

**Comment by yantosca** [comment]
Hi Liam, I am using GCC / Gfortran 9.2.0 and I have not come across this.   Which GCC version do you have?

**Comment by LiamBindle** [comment]
This is with 9.2.1. Have you tried the CMake and GNU Make builds?

**Comment by yantosca** [comment]
Yes, I have been using both.  But my version is 9.2.0 (spack-built).  I suspect this is some creep towards the newer F2008/F2018 standards in GCC that often leaves legacy code in the lurch.

I can try to implement a workaround.  If time permits, I can also see if they have gfortran 9.2.1 on spack and add that to my stack.

**Comment by LiamBindle** [comment]
To reproduce the compiler error you can do

```shell
# Pull and run penelope's GCC 9 line
docker pull liambindle/penelope:0.1.0-ubuntu16.04-gcc9-netcdf4.5.0-netcdff4.4.4
docker run -it liambindle/penelope:0.1.0-ubuntu16.04-gcc9-netcdf4.5.0-netcdff4.4.4
# We're in the container now...now configure the environment
source /init.rc
module load gcc/9
spack load netcdf
spack load netcdf-fortran
# Now we're ready to build GEOS-Chem
git clone https://github.com/geoschem/geos-chem.git
cd geos-chem
mkdir build
cd build
cmake .. -DRUNDIR=IGNORE -DRUNDIR_SIM=standard
make -j
```

**Comment by yantosca** [comment]
Ah I see the issue.   We have a DO loop where N is the loop variable, but in this code, N is an output argument:

https://github.com/geoschem/geos-chem/blob/601e919cf1163839d743daea0d835bde88a3de82/HEMCO/Core/hco_config_mod.F90#L991-L1005

The solution would be to rename that N in hco_charsplit to something else.  I can put that into dev/12.6.1.

**Comment by yantosca** [comment]
I have pushed https://github.com/geoschem/geos-chem/commit/c349b8a1d9a75e10bc837ac316b09814d4a14422, which resolves this issue.  We no longer use the same variable (N) to return output from HCO_CharSplit.

This build was verified with a full Azure DevOps build matrix:
https://dev.azure.com/geoschem/geos-chem/_build/results?buildId=3

**Comment by yantosca** [comment]
Now included in GC 12.6.1, released on 28 Oct 2019.
